pkg/server/plugin/nodeattestor/gcp/iit.go

```diff
@@ -0,0 +1,197 @@
+)
+
+type tokenKeyRetriever interface {
+	retrieveKey(token *jwt.Token) (interface{}, error)
+}
+
+type googlePublicKeyRetriever struct {
+	certificates map[string]*x509.Certificate
+	expirey      int64
+	mtx          *sync.Mutex
+}
+
+func (r *googlePublicKeyRetriever) retrieveKey(token *jwt.Token) (interface{}, error) {
+	if token.Header["kid"] == nil {
+		return nil, fmt.Errorf("Missing kid in identityToken header. Cannot verify token")
+	}
+	kid := token.Header["kid"].(string)
+	if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+		return nil, fmt.Errorf("Unexpected signing method: %v", token.Header["alg"])
+	}
+
+	if r.expirey == 0 || time.Now().Unix() > r.expirey {
+		r.mtx.Lock()
+		defer r.mtx.Unlock()
+		err := r.downloadCertificates()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return r.certificates[kid].PublicKey, nil
+}
+
+func (r *googlePublicKeyRetriever) downloadCertificates() error {
+	resp, err := http.Get(googleCertURL)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	bytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	var data map[string]string
+	err = json.Unmarshal(bytes, &data)
+
+	for k, v := range data {
+		block, _ := pem.Decode([]byte(v))
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return err
+		}
+		r.certificates[k] = cert
+	}
+	t, err := time.Parse("Mon, 2 Jan 2006 15:04:05 MST", resp.Header["Expires"][0])
+	r.expirey = t.Unix()
+	return nil
+}
+
+type IITAttestorConfig struct {
+	TrustDomain        string   `hcl:"trust_domain"`
+	ProjectIDWhitelist []string `hcl:"projectid_whitelist`
+}
+
+type IITAttestorPlugin struct {
+	trustDomain        string
+	projectIDWhitelist []string
+	tokenKeyRetriever  tokenKeyRetriever
+	mtx                *sync.Mutex
+}
+
+func (p *IITAttestorPlugin) spiffeID(gcpAccountID string, gcpInstanceID string) *url.URL {
+	spiffePath := path.Join("spire", "agent", pluginName, gcpAccountID, gcpInstanceID)
+	id := &url.URL{
+		Scheme: "spiffe",
+		Host:   p.trustDomain,
+		Path:   spiffePath,
+	}
+	return id
+}
+
+func (p *IITAttestorPlugin) Attest(ctx context.Context, req *nodeattestor.AttestRequest) (*nodeattestor.AttestResponse, error) {
+	if req == nil || req.GetAttestedData() == nil {
+		err := cgcp.AttestationStepError("retrieving the attested data", fmt.Errorf("AttestRequest or attestedData is nil"))
+		return &nodeattestor.AttestResponse{Valid: false}, err
+	}
+
+	identityToken := &cgcp.IdentityToken{}
+	_, err := jwt.ParseWithClaims(string(req.GetAttestedData().Data), identityToken, p.tokenKeyRetriever.retrieveKey)
+	if err != nil {
+		err = cgcp.AttestationStepError("parsing the identity token", err)
+		return &nodeattestor.AttestResponse{Valid: false}, err
+	}
+
+	if identityToken.Audience != audience {
+		err = cgcp.AttestationStepError("Audience claim in the token doesn't match the expected audience", err)
+		return &nodeattestor.AttestResponse{Valid: false}, err
+	}
+
+	if req.AttestedBefore {
```

Comment:
nit: moving this up near the top of the method will save some of the parsing and cryptographic validation overhead on tokens that have already been used